File: maui-sample/AccessCode/AccessCodeV2.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace MauiSample.AccessCode;

static public class AccessCodeV2
{
  private static RSA? _publicKey;

  public static void SetPublicKey(string jwkJson)
  {
    ArgumentException.ThrowIfNullOrWhiteSpace(jwkJson);

    using JsonDocument document = JsonDocument.Parse(jwkJson);
    JsonElement root = document.RootElement;

    if (root.GetProperty("kty").GetString() != "RSA")
    {
      throw new ArgumentException("Only RSA JWK keys are supported.", nameof(jwkJson));
    }

    if(!root.TryGetProperty("n", out JsonElement n) || !root.TryGetProperty("e", out JsonElement e))
    {
      throw new ArgumentException("n and e properties are required.", nameof(jwkJson));
    }
    
    if(n.GetString() is not string modulusString || e.GetString() is not string exponentString)
    {
      throw new ArgumentException("n and e properties must be strings.", nameof(jwkJson));
    }

    byte[] modulus = Base64UrlDecode(modulusString);
    byte[] exponent = Base64UrlDecode(exponentString);

    var rsa = RSA.Create();
    rsa.ImportParameters(new RSAParameters
    {
      Modulus = modulus,
      Exponent = exponent,
    });

    _publicKey?.Dispose();
    _publicKey = rsa;
  }

  private static byte[] Base64UrlDecode(string base64Url)
  {
    string base64 = base64Url.Replace('-', '+').Replace('_', '/');
    int padding = (4 - base64.Length % 4) % 4;
    return Convert.FromBase64String(base64.PadRight(base64.Length + padding, '='));
  }

  public static string Generate(Guid appID, DateTime utcTime)
  {
    if(_publicKey is null)
    {
      throw new InvalidOperationException("public key not set");
    }

    string lowercaseID = appID.ToString("D").ToLowerInvariant();
    // Format utcTime as an ISO8601 compliant string, like this: 2024-02-22T18:00:00Z
    string iso8601Time = utcTime.ToString("yyyy-MM-dd'T'HH:mm:ssK", CultureInfo.InvariantCulture);
    string plaintextAccessCode = lowercaseID + " " + iso8601Time;
    byte[] utf8Bytes = Encoding.UTF8.GetBytes(plaintextAccessCode);
    byte[] encryptedBytes = _publicKey.Encrypt(utf8Bytes, RSAEncryptionPadding.OaepSHA256);
    string base64String = Convert.ToBase64String(encryptedBytes);
    return base64String;
  }
}

File: ios-native-sample/ios-native-sample/AccessCode/AccessCodeV2.swift
import Foundation
import Security

enum AccessCodeV2 {
  nonisolated(unsafe) private static var publicKey: SecKey?

  private static func makeISO8601Formatter() -> ISO8601DateFormatter {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime]
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    return formatter
  }

  static func setPublicKey(_ jwkJson: String) throws {
    guard !jwkJson.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      throw AccessCodeV2Error.invalidJwk("JWK JSON must not be blank.")
    }

    guard let data = jwkJson.data(using: .utf8),
          let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
    else {
      throw AccessCodeV2Error.invalidJwk("Invalid JWK JSON.")
    }

    guard json["kty"] as? String == "RSA" else {
      throw AccessCodeV2Error.invalidJwk("Only RSA JWK keys are supported.")
    }

    guard let modulusString = json["n"] as? String,
          let exponentString = json["e"] as? String
    else {
      throw AccessCodeV2Error.invalidJwk("n and e properties are required.")
    }

    let modulus = try base64UrlDecode(modulusString)
    let exponent = try base64UrlDecode(exponentString)

    var error: Unmanaged<CFError>?
    let attributes: [String: Any] = [
      kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
      kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
    ]

    guard let key = SecKeyCreateWithData(
      rsaPublicKeyData(modulus: modulus, exponent: exponent) as CFData,
      attributes as CFDictionary,
      &error
    ) else {
      throw error?.takeRetainedValue() ?? AccessCodeV2Error.invalidJwk("Failed to create RSA public key.")
    }

    publicKey = key
  }

  static func generate(appId: UUID, utcTime: Date) throws -> String {
    guard let key = publicKey else {
      throw AccessCodeV2Error.publicKeyNotSet
    }

    let lowercaseId = appId.uuidString.lowercased()
    let iso8601Time = makeISO8601Formatter().string(from: utcTime)
    let plaintextAccessCode = "\(lowercaseId) \(iso8601Time)"
    let plaintextData = Data(plaintextAccessCode.utf8)

    var error: Unmanaged<CFError>?
    guard let encrypted = SecKeyCreateEncryptedData(
      key,
      .rsaEncryptionOAEPSHA256,
      plaintextData as CFData,
      &error
    ) as Data? else {
      throw error?.takeRetainedValue() ?? AccessCodeV2Error.encryptionFailed
    }

    return encrypted.base64EncodedString()
  }

  private static func base64UrlDecode(_ base64Url: String) throws -> Data {
    var base64 = base64Url
      .replacingOccurrences(of: "-", with: "+")
      .replacingOccurrences(of: "_", with: "/")
    let padding = (4 - base64.count % 4) % 4
    base64 += String(repeating: "=", count: padding)
    guard let data = Data(base64Encoded: base64) else {
      throw AccessCodeV2Error.invalidJwk("Invalid base64url encoding.")
    }
    return data
  }

  /// Builds DER-encoded RSAPublicKey (PKCS#1) for SecKeyCreateWithData.
  private static func rsaPublicKeyData(modulus: Data, exponent: Data) -> Data {
    let modulusEncoded = derInteger(modulus)
    let exponentEncoded = derInteger(exponent)
    let sequenceBody = modulusEncoded + exponentEncoded
    return derSequence(sequenceBody)
  }

  private static func derInteger(_ value: Data) -> Data {
    var bytes = value
    if bytes.first == 0x00 || (bytes.first ?? 0) >= 0x80 {
      bytes = Data([0x00]) + bytes
    }
    return Data([0x02]) + derLength(bytes.count) + bytes
  }

  private static func derSequence(_ body: Data) -> Data {
    Data([0x30]) + derLength(body.count) + body
  }

  private static func derLength(_ length: Int) -> Data {
    if length < 128 {
      return Data([UInt8(length)])
    }
    var value = length
    var bytes: [UInt8] = []
    while value > 0 {
      bytes.insert(UInt8(value & 0xFF), at: 0)
      value >>= 8
    }
    return Data([UInt8(0x80 | bytes.count)] + bytes)
  }
}

enum AccessCodeV2Error: Error, LocalizedError {
  case publicKeyNotSet
  case encryptionFailed
  case invalidJwk(String)

  var errorDescription: String? {
    switch self {
    case .publicKeyNotSet:
      return "Public key not set."
    case .encryptionFailed:
      return "RSA encryption failed."
    case .invalidJwk(let message):
      return message
    }
  }
}
